crates/subspace-service/src/mmr/sync.rs

```diff
@@ -137,18 +190,46 @@ where
                         // Save the MMR-nodes from response to the local storage
                         'data: for (position, data) in response.mmr_data.iter() {
                             // Ensure continuous sync
-                            if *position == starting_position {
-                                let canon_key = get_offchain_key((*position).into());
-                                offchain_db.local_storage_set(
-                                    StorageKind::PERSISTENT,
-                                    &canon_key,
-                                    data,
+                            if *position != starting_position {
+                                debug!(
+                                    ?peer_info,
+                                    %starting_position,
+                                    %position,
+                                    "MMR sync error: incorrect starting position."
                                 );
 
-                                starting_position += 1;
-                            } else {
-                                debug!("MMR-sync gap detected={peer_id}, position={position}",);
-                                break 'data; // We don't support gaps in MMR data
+                                continue 'peers;
+                            }
+
+                            let node = decode_mmr_data(data);
+
+                            let node = match node {
+                                Ok(node) => node,
+                                Err(err) => {
+                                    debug!(?peer_info, ?err, %position, "Can't decode MMR data received from the peer.");
+
+                                    continue 'peers;
+                                }
+                            };
+
+                            if matches!(node, Node::Data(_)) {
+                                if let Err(err) = mmr.push(node) {
```

Comment:
So you're doing pushes, but aren't you supposed to call `mmr.commit()` at the end? It is not documented, but according to internals doesn't actually store anything immediately.

Which is also a bit concerning because it doesn't guarantee atomic writes it seems, requiring re-sync in case of interruption (which will violate your "empty DB" assumption below).